### PR TITLE
[IA-267] Support messages API with pagination

### PR DESCRIPTION
**Pull request by Undermaken**
This PR includes all those changes to support pagination of messages list, as described in this doc
https://pagopa.atlassian.net/wiki/spaces/IC/pages/358909575/Messaggi+-+Paginazione

### CODE REVIEW

File: src/routers/message.ts [hunk 1]
```diff
@@ -166,17 +167,106 @@ const createMessages = () => {
...
+    minimumId: req.query.minimum_id
+  });
+  if (paginatedQuery.isLeft()) {
+    // bad request
+    res.sendStatus(400);
+    return;
+  }
+
+  const params = paginatedQuery.value;
+  // order messages by creation date (desc)
```
Comment (pietro909) [inline src/routers/message.ts:214]: I understand that they are ordered [by ID](https://pagopa.atlassian.net/wiki/spaces/IC/pages/358909575/Messaggi+-+Paginazione#La-soluzione)

Comment (Undermaken) [reply src/routers/message.ts:214]: Yes, and actually it is!
This implementation is a simplification of what IO backend does. 
IO backend works with a cosmoDB and a storage to retrieve messages.

Messages are immutable so backend orders them by ID and that's ensure they are ordered by creation date:
`ID2 > ID1` -> `creationDate2 > creationDate1`. They use GUID given by cosmosDB to implement this ordering (I guess)

In our scenario we have only the string ID (could represent a number or not) of each message and I think that it could be dangerous trying an ordering using strings. So I point at the expected result and keep it simple: the list should be ordered by creation date

Since your point spawns some doubts about my reasons, I asked to Micheal to confirm them
Schermata 2021-10-01 alle 09 10 06

File: src/routers/message.ts [hunk 2]
```diff
@@ -166,17 +167,106 @@ const createMessages = () => {
...
+    const endIndex = orderedList.findIndex(m => m.id === params.minimumId);
+    if (endIndex <= 0) {
+      res.json({ items: [] });
+      return;
+    }
+    indexes = {
+      startIndex: Math.max(0, endIndex - (1 + params.pageSize!)),
+      endIndex
+    };
+  }
```
Comment (pietro909) [inline src/routers/message.ts:257]: I find the large and nested `if/else` block a tad too convoluted. There some exit conditions (the `return`) whithin it that make it hard to follow.

You may initialize `indexes = null` and then match against it down the line.

Example (using `switch` just because I like it 🤓 ) :

```typescript
let indexes = null

switch ([typeof minimumId,typeof maximumId]) {

  // both boundaries
  case ["string", "string"]: {
    // ...more checks
      minimumId.toString()
      indexes = {
        startIndex: startIndex + 1,
        endIndex
      };
  }

  // only lower bound
  case ["string", undefined]: {
    // ...more checks
    indexes = {
      startIndex: Math.max(0, endIndex - (1 + params.pageSize!)),
      endIndex
    };
  }

  // only upper bound
  case [undefined,"string"]: {
    // ...more checks
    indexes = {
      startIndex: Math.max(0, endIndex - (1 + params.pageSize!)),
      endIndex
    };
  }
}

// exit condition
if (indexes === null) {
  res.json({ items: [] });
  return;
}

// from here, work with indexes
```

Comment (Undermaken) [reply src/routers/message.ts:280]: Thank you, It was a sentiment I had but I want to ignore 🙈 
Your hint it's surely better. In this case I see a clear _pattern-matching_ so I introduced I used [ts-patterns](https://github.com/gvergnaud/ts-pattern)

What do you think? Could it be a good refactoring? 
https://github.com/pagopa/io-dev-api-server/pull/79/commits/7c4d18b8c5a06d8cf58ac21ed4182fcb1e556f04

Comment (Undermaken) [reply src/routers/message.ts:274]: I did refactoring inverting some condition and remove useless `return undefined`
https://github.com/pagopa/io-dev-api-server/pull/79/commits/a69a770d79bf14087ef804d073740b18519f0184
